buildtools/maven/maven.go

```diff
@@ -59,53 +72,141 @@ func (m *Maven) Compile(dir string) error {
 	return err
 }
 
-func Modules(dir string) ([]string, error) {
+// ResolveManifestFromBuildTarget tries to determine what buildTarget is supposed to be and then reads the POM
+// manifest file pointed to by buildTarget if it is a path to such a file or module.
+func (m *Maven) ResolveManifestFromBuildTarget(buildTarget string) (*Manifest, error) {
+	var pomFile string
+	if !strings.HasSuffix(buildTarget, ".xml") {
+		pomFile = filepath.Join(buildTarget, "pom.xml")
+	} else {
+		pomFile = buildTarget
+	}
+	if ok, _ := files.Exists(pomFile); !ok {
+		if strings.Count(buildTarget, ":") == 1 {
+			// This is likely a module ID.
+			return nil, fmt.Errorf("build target %q appears to be a module identifier", buildTarget)
+		}
+		return nil, fmt.Errorf("module manfist file %q does not exist", buildTarget)
+	}
+
 	var pom Manifest
-	err := files.ReadXML(&pom, filepath.Join(dir, "pom.xml"))
-	if err != nil {
-		return nil, errors.Wrap(err, "could not read POM file")
+	if err := files.ReadXML(&pom, pomFile); err != nil {
+		return nil, err
 	}
+	if m.manifestCache == nil {
+		m.manifestCache = make(map[string]*Manifest)
+	}
+	m.manifestCache[buildTarget] = &pom
+	return &pom, nil
+}
+
+// A MvnModule can identify a Maven project with the target path.
+type MvnModule struct {
+	// Name is taken from the module's POM file.
+	Name string
+
+	// Target is the relative path from the root of the FOSSA project to the module's manifest file or to
+	// the directory containing the module's "pom.xml". The target may name a file other than "pom.xml" and
+	// it may even be the groupId:artifactId string identifying the "Maven project".
+	Target string
+
+	// Dir is the relative path from the root of the FOSSA project to the module.
+	Dir string
+}
+
+// Modules returns a list specifying the Maven module at path, which may name a file or directory, and all the
+// Maven modules nested below it. The Target field of each MvnModule is set to the directory or manifest file
+// describing the module. The visited manifest files are listed in the checked map.
+func Modules(path string, baseDir string, checked map[string]bool) ([]MvnModule, error) {
+	dir := path
+	pomFile := path
+	if strings.HasSuffix(path, ".xml") {
+		// We have the manifest file but still need its directory path.
+		dir = filepath.Dir(path)
```

Comment:
Probably a big edge case, but what if the user defines a  custom Pom manifest file without an xml suffix? Is there maybe a more reliable check we could do here?

Reply:
fixed